packages/compass-components/src/components/resizeable-sidebar.tsx

```diff
@@ -57,24 +57,68 @@ const containerStylesLight = css({
   backgroundColor: 'var(--bg-color)',
 });
 
-const initialSidebarWidth = spacing[6] * 4 - spacing[1]; // 252px
-const minSidebarWidth = spacing[4] * 9; // 216px
-
 const ResizableSidebar = ({
-  initialWidth = initialSidebarWidth,
-  minWidth = minSidebarWidth,
+  collapsable = false,
+  expanded = true,
+  setExpanded = () => {
+    return;
+  },
+  initialWidth = spacing[6] * 4,
+  minWidth = 210,
+  collapsedWidth = 48,
   children,
 }: {
+  collapsable?: boolean;
+  expanded?: boolean;
+  setExpanded?: (isExpaned: boolean) => void;
   initialWidth?: number;
   minWidth?: number;
+  collapsedWidth?: number;
   children: JSX.Element;
 }): JSX.Element => {
   const [width, setWidth] = useState(initialWidth);
+  const [prevWidth, setPrevWidth] = useState(initialWidth);
 
   const getMaxSidebarWidth = useCallback(() => {
     return Math.max(minWidth, window.innerWidth - 100);
   }, [minWidth]);
 
+  // Apply bounds to the sidebar width when resizing to ensure it's always
+  // visible and usable to the user.
+  const boundSidebarWidth = useCallback(
+    (attemptedWidth: number) => {
+      const maxWidth = getMaxSidebarWidth();
+
+      return Math.min(maxWidth, Math.max(minWidth, attemptedWidth));
+    },
+    [getMaxSidebarWidth, minWidth]
+  );
+
+  const updateWidth = useCallback(
+    (newWidth: number) => {
+      setWidth(newWidth);
+
+      const expectedExpanded = width > minWidth || !collapsable;
+      if (expanded !== expectedExpanded) {
+        setExpanded(expectedExpanded);
+      }
+
+      // Keep the most recent expanded width in case the sidebar suddenly gets
+      // collapsed (via a button or keyboard shortcut or similar) so that if it
+      // gets similarly expanded later we can restore the width.
+      if (expectedExpanded) {
+        setPrevWidth(newWidth);
+      }
+    },
+    [collapsable, expanded, minWidth, setExpanded, width]
+  );
+
+  const renderedWidth = expanded ? boundSidebarWidth(width) : collapsedWidth;
+
+  if (expanded && width === collapsedWidth) {
+    setWidth(prevWidth);
+  }
```

Comment:
```suggestion
  useEffect(() => {
    if (expanded && width === collapsedWidth) {
      setWidth(prevWidth);
    }
  }, [ setWidth, expanded, width, collapsedWidth ]);
```
Could this benefit from a `useEffect`?

Reply:
I changed it, but thinking about it more it might be the rare case where we want something to happen _before_ render and maybe even in place of it rather than after. Or at least not render until the width's been updated. As this will likely cause multiple renders. Oh well.

Reply:
There's a `useLayoutEffect` hook that's maybe what you're looking for:
https://reactjs.org/docs/hooks-reference.html#uselayouteffect
It runs before html renders.